### PR TITLE
Collect metrics for notification delivery SLIs

### PR DESCRIPTION
**Pull request by jacobwinch**
## What does this change?

This PR updates the SLO monitoring code (first introduced via https://github.com/guardian/mobile-n10n/pull/719) to capture CloudWatch metrics related to notification delivery times.

The metrics captured here allow us to [calculate various SLIs](https://metrics.gutools.co.uk/goto/lw3X88FVk) related to notification delivery performance:

image
_N.B. the SLI values here are just examples based on test data_

We can also get platform specific SLIs (this seems useful given the potential for performance differences between Firebase and APNS):

image
_N.B. the SLI values here are just examples based on test data_

image
_N.B. the SLI values here are just examples based on test data_

## How to test

I've tested this by deploying to `CODE` and adding a [small hack](https://github.com/guardian/mobile-n10n/commit/c6547742484ce00e63989e72af650b991b13c281) which allows me to query production data. 

## How can we measure success?

This will allow us to visualise some useful SLIs. In the future, we could also consider using these metrics as the basis for alerts.

## Have we considered potential risks?

* The current SLI calculations (in Grafana, not in this codebase) use the total number of deliveries metric rather than a (currently absent) _expected_ number of deliveries metric. We might want to amend this later as using the latter would allow us to pick up cases where we had a coverage problem as well as a performance problem (currently if we only deliver 1 notification but we deliver it very quickly, the SLI value will be 100%!).
* @DavidLawes mentioned that the MSS team is using the structured logging produced by this Lambda; I have left the original logging in place (although the calculation is slightly different) to avoid breaking anything.

### CODE REVIEW

File: slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala [hunk 1]
```diff
@@ -47,41 +92,76 @@ object SloMonitor {
...
+       |  END) AS less_than_240,
+       |  SUM(CASE
+       |    WHEN DATE_DIFF('second', from_iso8601_timestamp('$sentTime'), received_timestamp) < 270 THEN 1
+       |    ELSE 0
+       |  END) AS less_than_270,
+       |  SUM(CASE
+       |    WHEN DATE_DIFF('second', from_iso8601_timestamp('$sentTime'), received_timestamp) < 300 THEN 1
+       |    ELSE 0
+       |  END) AS less_than_300,
+       |  COUNT(*) AS total_deliveries
```
Comment (jacobwinch) [inline slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala:137]: N.B. that `total_deliveries` really means `total_deliveries_when_this_query_was_run` - in practice that means the number of deliveries [in the first 10 minutes](https://github.com/guardian/mobile-n10n/blob/63dd9b0ca1b3f1020cc65c372c0c72fd1cb31403/notification/app/notification/services/SloTrackingSender.scala#L18), at present.

File: slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala [hunk 2]
```diff
@@ -47,41 +92,76 @@ object SloMonitor {
...
-          logger.info(Map(
-            "notificationId" -> notificationId,
-            "deliveriesWithin2mins" -> actualDeliveries
-          ),
-          s"Notifications delivered within 120 seconds was $actualDeliveries")
-        })
-      )
+    val result = Athena.startQuery(Query("notifications", query, s"s3://aws-mobile-event-logs-${this.stage.toLowerCase()}/athena/slo-monitoring")).flatMap {
+      Athena.fetchQueryResponse(_, rows => {
+        val androidDeliveries = rows.head
```
Comment (jacobwinch) [inline slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala:154]: The response data looks like this:

```
"platform","less_than_30","less_than_60","less_than_90","less_than_120","less_than_150","less_than_180","less_than_210","less_than_240","less_than_270","less_than_300","total_deliveries"
"android","424","432","436","437","439","441","441","441","441","441","441"
"ios","701","702","703","704","704","704","704","704","704","704","704"
```

It looks like the `GROUP BY` is sorting alphabetically by default but perhaps we should add some validation or something here to ensure that the query results look roughly how we expect (e.g. there should be X rows, Y columns, one row starting with `android` and one with `ios` etc.)

File: slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala (hunk 1 above)
Comment (DavidLawes) [reply slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala:137]: I think this is a good enough for now. I know that we've been keen to explore the number of missing/lost deliveries too, so at some point in the future we may want to update how we define the "total".

File: slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala (hunk 2 above)
Comment (DavidLawes) [reply slomonitor/src/main/scala/com.gu.notifications.slos/SloMonitor.scala:154]: Great idea to separate this by platform! I guess we could use metric math in the console to get an aggregated view of the 90in2 (if we ever wanted to do this)?

**Review by DavidLawes — APPROVED** [review]
Nice! This should help us get better insights more easily, thanks so much :)

As a side note, in our notification worker lambdas we've temporarily reverted to sending all metrics via the put metric api (instead of using embedded metric format) - this was to enable us to test the stack when running on ec2s instead of lambdas. Possibly when/if we revert that change we might experience the same logging issue you've come across here, so thanks for flagging it.

**Comment by jacobwinch** [comment]
> Great idea to separate this by platform!

It was inspired by the excellent presentation yesterday afternoon!

>  I guess we could use metric math in the console to get an aggregated view of the 90in2 (if we ever wanted to do this)?

Yes exactly, here's an example of that calculation in Grafana:

image

`((iosLessThan120 + androidLessThan120) / (iosTotalDeliveries + androidTotalDeliveries)) * 100`

**Comment by jacobwinch** [comment]
These metrics are coming through now; we can see the results so far in this [dashboard](https://metrics.gutools.co.uk/goto/NVfvdXK4k).

**Pull request merged**